overlord/snapshotstate/backend/backend.go

```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
 
 // Commit will commit a given transaction
 func (t *importTransaction) Commit() error {
-	if err := os.Remove(importInProgressFilepath(t.id)); err != nil {
+	t.committed = true
+	return t.unlock()
+}
+
+func (t *importTransaction) lock() error {
+	flock, err := osutil.NewFileLock(t.flockPath)
+	if err != nil {
 		return err
 	}
-	t.committed = true
+	if err := flock.Lock(); err != nil {
+		return err
+	}
+	t.flock = flock
 	return nil
 }
 
+func (t *importTransaction) unlock() error {
+	if t.flock == nil {
+		return nil
+	}
+	if err := t.flock.Close(); err != nil {
+		return err
+	}
+	if err := os.Remove(t.flockPath); err != nil {
+		return err
+	}
+	t.flock = nil
+	return nil
+}
+
+// CleanupAbandondedImports will clean any import that is in progress.
+// This is meant to be call at startup of snapd before any real imports
+// happen.
+//
+// The amount of snapshots cleaned is returned or an error
```

Comment:
It seems to actually clean as much as possible and return a number of cleaned snapshots *and* an error (if there was one), but see comment below.